overlord/ifacestate/apparmorprompting/noticebackend.go

```diff
@@ -0,0 +1,619 @@
+	// will be able to release the lock.
+	ntb.cond = sync.NewCond(ntb.rwmu.RLocker())
+	if err := ntb.load(now); err != nil {
+		return nil, err
+	}
+	return ntb, nil
+}
+
+// addNotice records an occurrence of a notice with the specified user ID, a
+// key equal to the given prompt/rule ID, and the given data, with notice ID
+// and type derived from the receiver.
+func (ntb *noticeTypeBackend) addNotice(userID uint32, id prompting.IDType, data map[string]string) error {
+	ntb.rwmu.Lock()
+	defer ntb.rwmu.Unlock()
+	key := id.String()
+	noticeID := fmt.Sprintf("%s-%s", ntb.namespace, key)
+
+	userNotices, existingNotice, existing, existingIndex, err := ntb.searchExistingNotices(userID, noticeID)
+	if err != nil {
+		return err
+	}
+
+	// Now that errors can't occur (other than save error), get a new unique
+	// timestamp from the state, which will bump the state's noticeLastTimestamp
+	timestamp := ntb.nextNoticeTimestamp()
+
+	// Check if any notices have expired relative to the new timestamp.
+	// Since they're sorted, as soon as we see a non-expired notice, bail out.
+	// Do this before potentially calling Reoccur on the existing notice, so we
+	// see its original timestamp in the sorted slice.
+	expiredCount := 0
+	for _, n := range userNotices {
+		if !n.Expired(timestamp) {
+			break
+		}
+		expiredCount++
+	}
+
+	var newNotice *state.Notice
+	if existing && !existingNotice.Expired(timestamp) {
+		newNotice = existingNotice.DeepCopy()
+		newNotice.Reoccur(timestamp, data, 0)
+	} else {
+		newNotice = state.NewNotice(noticeID, &userID, ntb.noticeType, key, timestamp, data, 0, defaultExpireAfter)
+	}
+
+	newUserNotices := assembleNewUserNotices(userNotices, expiredCount, newNotice, existing, existingIndex)
+
+	ntb.userNotices[userID] = newUserNotices
+	ntb.idToNotice[noticeID] = newNotice
+
+	if err := ntb.save(); err != nil {
+		ntb.userNotices[userID] = userNotices
+		if existing {
+			ntb.idToNotice[noticeID] = existingNotice
+		} else {
+			delete(ntb.idToNotice, noticeID)
+		}
+		return fmt.Errorf("cannot add notice to prompting %s backend: %w", ntb.noticeType, err)
+	}
+
+	// Now that we've successfully saved, delete the expired notices
+	for _, expiredNotice := range userNotices[:expiredCount] {
+		delete(ntb.idToNotice, expiredNotice.ID())
+	}
+
+	ntb.cond.Broadcast()
+
+	return nil
+}
+
+// searchExistingNotice looks up the list of existing notices for the given
+// userID and checks whether a notice with the given noticeID already exists.
+//
+// Returns the slice of existing notices for the given userID. If the notice
+// does exist, a pointer to it is returned, along with an existing boolean of
+// true and the index at which it occurs in the userNotices slice.
+//
+// The caller must ensure that the backend mutex is locked.
+func (ntb *noticeTypeBackend) searchExistingNotices(userID uint32, noticeID string) (userNotices []*state.Notice, notice *state.Notice, existing bool, existingIndex int, err error) {
```

Comment:
given that this is an internal helper and used in one place I would just drop existing and use notice != nil in the caller